### PR TITLE
Jonathan Yuan Bootcamp

### PR DESCRIPTION
**Pull request by JonathanYuan23**

### CODE REVIEW

File: modules/bootcamp/detect_landing_pad.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 """
 import pathlib
 
+import cv2
```
Comment (Xierumeng) [inline modules/bootcamp/detect_landing_pad.py:8]: Do not include additional imports!

File: modules/bootcamp/detect_landing_pad.py [hunk 2]
```diff
@@ -86,32 +87,50 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
 
         # Get the xyxy boxes list from the Boxes object in the Result object
-        boxes_xyxy = ...
+        boxes_xyxy = boxes.xyxy
 
         # Detach the xyxy boxes to make a copy,
         # move the copy into CPU space,
         # and convert to a numpy array
-        boxes_cpu = ...
+        boxes_cpu = boxes_xyxy.cpu().numpy()
```
Comment (Xierumeng) [inline modules/bootcamp/detect_landing_pad.py:105]: Prefer to detach before moving the the CPU.

File: modules/bootcamp/detect_landing_pad.py [hunk 3]
```diff
@@ -86,32 +87,50 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
         # Get the Result object
-        prediction = ...
+        prediction = predictions[0]
+        boxes = prediction.boxes
 
-        # Plot the annotated image from the Result object
-        # Include the confidence value
-        image_annotated = ...
+        confidences = boxes.conf
+        confidences_cpu = confidences.cpu().numpy()
```
Comment (Xierumeng) [inline modules/bootcamp/detect_landing_pad.py:97]: Not required.

File: modules/bootcamp/detect_landing_pad.py [hunk 4]
```diff
@@ -86,32 +87,50 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
+
+        for i in range(0, boxes_cpu.shape[0]):
+            box = tuple(boxes_cpu[i])
+            x1, y1, x2, y2 = int(box[0]), int(box[1]), int(box[2]), int(box[3])
+
+            image_annotated = cv2.rectangle(image_annotated, (x1, y1), (x2, y2), (36,255,12), 1)
+
+            confidence = confidences_cpu[i]
+
+            image_annotated = cv2.putText(image_annotated, str(confidence), (x1, y1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (36,255,12), 2)
```
Comment (Xierumeng) [inline modules/bootcamp/detect_landing_pad.py:120]: There is an Ultralytics method that will do this for you, use that instead.

File: modules/bootcamp/detect_landing_pad.py [hunk 5]
```diff
@@ -86,32 +87,50 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
         # Loop over the boxes list and create a list of bounding boxes
         bounding_boxes = []
         # Hint: .shape gets the dimensions of the numpy array
-        # for i in range(0, ...):
+        for i in range(0, boxes_cpu.shape[0]):
             # Create BoundingBox object and append to list
-            # result, box = ...
+            bounds = boxes_cpu[i]
+            result, box = bounding_box.BoundingBox.create(bounds)
+            assert(result)
```
Comment (Xierumeng) [inline modules/bootcamp/detect_landing_pad.py:129]: Avoid raising exceptions and return an empty list instead.

File: modules/bootcamp/decision_simple_waypoint.py [hunk 6]
```diff
@@ -6,6 +6,7 @@
 # Disable for bootcamp use
 # pylint: disable=unused-import
 
+import math
```
Comment (Xierumeng) [inline modules/bootcamp/decision_simple_waypoint.py:9]: Do not include additional imports.

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 7]
```diff
@@ -6,6 +6,7 @@
 # Disable for bootcamp use
 # pylint: disable=unused-import
 
+import math
```
Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:9]: No additional imports.

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 8]
```diff
@@ -68,10 +69,59 @@ def run(self,
...
+        if report.status == drone_status.DroneStatus.HALTED:
+            if self.timeline == "initial":
+                x2, y2 = self.waypoint.location_x, self.waypoint.location_y
+                relative_x = x2 - x1
+                relative_y = y2 - y1
+                command = commands.Command.create_set_relative_destination_command(relative_x, relative_y)
+
+                self.timeline = "initial-moving"
+            elif self.timeline == "initial-moving":
+                closest_x, closest_y = landing_pad_locations[0].location_x, landing_pad_locations[0].location_y
```
Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:86]: While the bootcamp specifies at least 1 landing pad, what happens if there were no landing pads? (You do not need to change anything here, just something to think about).

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 9]
```diff
@@ -68,10 +69,59 @@ def run(self,
...
+                        closest_x = new_x
+                        closest_y = new_y
+
+                relative_x = closest_x - x1
+                relative_y = closest_y - y1
+                command = commands.Command.create_set_relative_destination_command(relative_x, relative_y)
+
+                self.timeline = "waypoint-moving"
+                self.landing_x = closest_x
+                self.landing_y = closest_y
```
Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:104]: Declare these in the constructor.

File: modules/bootcamp/detect_landing_pad.py (hunk 2 above)
Comment (JonathanYuan23) [reply modules/bootcamp/detect_landing_pad.py:105]: Sorry could you elaborate 😅? I thought `boxes_xyxy = boxes.xyxy` was detaching the xyxy boxes.

Comment (Xierumeng) [reply modules/bootcamp/detect_landing_pad.py:101]: Pytorch tensors have a specific `detach()` method.

File: modules/bootcamp/detect_landing_pad.py [hunk 10]
```diff
@@ -86,32 +86,39 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
         # Detach the xyxy boxes to make a copy,
         # move the copy into CPU space,
         # and convert to a numpy array
-        boxes_cpu = ...
+        boxes_cpu = boxes_xyxy.cpu().numpy()
+
+        # Plot the annotated image from the Result object
+        # Include the confidence value
+
+        image_annotated = prediction.plot()
```
Comment (Xierumeng) [inline modules/bootcamp/detect_landing_pad.py:106]: Do the output images have the confidence values?

File: modules/bootcamp/decision_simple_waypoint.py [hunk 11]
```diff
@@ -17,6 +16,8 @@
 # Disable for bootcamp use
 # pylint: disable=unused-argument,line-too-long
 
+def distance(x1: float, y1: float, x2: float, y2: float) -> float:
```
Comment (Xierumeng) [inline modules/bootcamp/decision_simple_waypoint.py:19]: Good use of type hinting!

File: modules/bootcamp/decision_simple_waypoint.py [hunk 12]
```diff
@@ -17,6 +16,8 @@
...
+    return (x2 - x1)**2 + (y2 - y1)**2
```
Comment (Xierumeng) [inline modules/bootcamp/decision_simple_waypoint.py:20]: Good use of avoiding square roots!

File: modules/bootcamp/decision_simple_waypoint.py [hunk 13]
```diff
@@ -68,10 +69,27 @@ def run(self,
...
+
         # Do something based on the report and the state of this class...
+        if report.status == drone_status.DroneStatus.HALTED:
+            if self.timeline == "initial":
+                relative_x = x2 - x1
+                relative_y = y2 - y1
+                command = commands.Command.create_set_relative_destination_command(relative_x, relative_y)
+
+                self.timeline = "moving"
+            else:
```
Comment (Xierumeng) [inline modules/bootcamp/decision_simple_waypoint.py:83]: What if the drone was allowed to halt in the middle of a movement? (You do not need to change anything here, just something to think about).

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 14]
```diff
@@ -68,10 +71,59 @@ def run(self,
         # ↓ BOOTCAMPERS MODIFY BELOW THIS COMMENT ↓
         # ============
 
-        # Do something based on the report and the state of this class...
+        # print(landing_pad_locations)
```
Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:74]: Do not leave commented code in, remove.

File: modules/bootcamp/detect_landing_pad.py (hunk 10 above)
Comment (JonathanYuan23) [reply modules/bootcamp/detect_landing_pad.py:106]: Yes it does.

image

File: modules/bootcamp/detect_landing_pad.py [hunk 15]
```diff
@@ -86,32 +86,39 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
             # Create BoundingBox object and append to list
-            # result, box = ...
-
-        # Remove this when done
-        raise NotImplementedError
+            bounds = boxes_cpu[i]
+            result, box = bounding_box.BoundingBox.create(bounds)
+            
+            if not result:
+                [], image_annotated
```
Comment (Xierumeng) [inline modules/bootcamp/detect_landing_pad.py:117]: Does this return?

**Review by Xierumeng — COMMENTED** [review]
LGTM.

**Pull request closed without merging**